### ISSUE TITLE
Error: [2057] when fetching remote table at non-HEAD commit following schema change using as of

### ISSUE THREAD

**Issue by n8layman** [dolthub/dolt#3583]
Trying to fetch non-HEAD commits produces an error if schema has been changed when using the [doltr](https://github.com/ecohealthalliance/doltr) R package.

`Error: The number of parameters in bound buffers differs from number of columns in resultset [2057]`

doltr issue [#46](https://github.com/ecohealthalliance/doltr/issues/46)

I'll post dolt server logs this evening

**Comment by timsehn** [comment]
We were able to reproduce. 

We think it is a bug related to the calculation of the return size of the query not matching what Dolt actually sends back. The R or MariaDB client uses this information and errors on mismatch. Other clients we've tested don't seem to do this or we just have not tested them thoroughly enough.

**Comment by timsehn** [comment]
Seems only reproducible in R. Here is python with the `mariadb` connector.
```
import mariadb
import sys

def main():
    user = 'root'
    port = 3306
    db = 'test_as_of'

    connection = mariadb.connect(user=user,
                                 host="127.0.0.1",
                                 port=port,
                                 database=db)

    query = "select * from test as of 'HEAD^'"

    cursor = connection.cursor()
    cursor.execute(query)

    results = cursor.fetchall()
    print(results)

    cursor.close()

    connection.close()

    exit(0)

main()
```
```bash
dolt $ python test.py 
[(0, 0), (1, 1)]
```

**Comment by noamross** [comment]
@klrmlr (RMariaDB maintainer) any thoughts on what triggers this?

**Comment by n8layman** [comment]
Thanks for narrowing it down! You were able to reproduce the error though?

**Comment by timsehn** [comment]
This is the only place in the C connector that error can happen:
```
https://github.com/mariadb-corporation/mariadb-connector-c/blob/a9dcf924544ebf9491f638069587b1dc4cfb66bb/libmariadb/mariadb_stmt.c#L2067
```

**Comment by timsehn** [comment]
> Thanks for narrowing it down! You were able to reproduce the error though?

Yes. We were able to reproduce in the R client. Trying to get it to fail in other clients.

**Comment by fulghum** [comment]
We've been experimenting with our repro in RStudio and looking through the Dolt sql-serve trace logs. This same query seems to execute differently inside Dolt depending on whether it's from a MySQL connection or a MariaDB connection. 

In a MySQL connection, here's what we see being returned: 
```
2022-06-09T17:53:23-07:00 DEBUG [conn 5] returning result &{[name:"measurement_id" type:INT64 charset:33 name:"sepal_length_cm" type:FLOAT64 charset:33 name:"sepal_width_cm" type:FLOAT64 charset:33 name:"petal_length_cm" type:FLOAT64 charset:33 name:"petal_width_cm" type:FLOAT64 charset:33 name:"class" type:TEXT charset:33 name:"stem_length_cm" type:FLOAT64 charset:33] 1 0  [[INT64(1) FLOAT64(5.1) FLOAT64(3.5) FLOAT64(1.4) FLOAT64(0.2) TEXT("setosa") FLOAT64(97.05147377090773)]] <nil>} {connectTime=2022-06-09T17:53:13-07:00, connectionDb=iris_flower_dataset, query=SELECT * from classified_measurements as of 'idvhbfo2qua9ir6iifj1hipa88br21r7' LIMIT 1;}
```

Note that you can see all the fields being correctly returned. 

And here's what we see for the MariaDB connection:
```
2022-06-09T17:54:56-07:00 DEBUG [conn 6] returning result &{[name:"measurement_id" type:INT64 charset:33 name:"sepal_length_cm" type:FLOAT64 charset:33 name:"sepal_width_cm" type:FLOAT64 charset:33 name:"petal_length_cm" type:FLOAT64 charset:33 name:"petal_width_cm" type:FLOAT64 charset:33 name:"class" type:TEXT charset:33] 1 0  [[INT64(1) FLOAT64(5.1) FLOAT64(3.5) FLOAT64(1.4) FLOAT64(0.2) TEXT("setosa")]] <nil>} {connectTime=2022-06-09T17:54:35-07:00, connectionDb=iris_flower_dataset, query=SELECT * from classified_measurements as of 'idvhbfo2qua9ir6iifj1hipa88br21r7' LIMIT 1;}
```

You can see that Dolt is not sending back the last field (`stem_length_cm`) over the MariaDB connection. That field was removed from the table in a commit between `idvhbfo2qua9ir6iifj1hipa88br21r7` and `HEAD`. There are also some other log statements that seem to show the statement is being optimized and rewritten **only** in the MariaDB connection, and it seems likely that there is a bug in that statement optimization logic that is not honoring the `as of` field. 

We're continuing to dig into this to determine why the Dolt is executing the query differently over the MariaDB connection and have a few ideas of things to try next.

**Comment by n8layman** [comment]
Thanks for looking into it! 

I'm no expert at SQL server connectors but here's a relevant chunk from the logs I have access to. In particular I was wondering about the `plan after re-opt: Exchange └─ Table(test)` bit. Do you know what that is all about? Is MariaDB retrying the connection but substituting the table name, in this case `test`, without the 'as of' part of the query?

{connectTime=2022-06-09T18:05:44-07:00, connectionDb=test_db, query=SELECT * FROM test as of 'HEAD^1'}
2022-06-09T18:05:49-07:00 TRACE[conn 2] beginning new transaction
{connectTime=2022-06-09T18:05:44-07:00, connectionDb=test_db, query=SELECT * FROM test as of 'HEAD^1'}
2022-06-09T18:05:49-07:00 TRACE[conn 2] optimizing prepared plan for query: SELECT * FROM test as of 'HEAD^1'
{connectTime=2022-06-09T18:05:44-07:00, connectionDb=test_db, query=SELECT * FROM test as of 'HEAD^1'}
2022-06-09T18:05:49-07:00 TRACE[conn 2] plan before re-opt: Projected table access on [a b] └─ Table(test)
{connectTime=2022-06-09T18:05:44-07:00, connectionDb=test_db, query=SELECT * FROM test as of 'HEAD^1'}
2022-06-09T18:05:49-07:00 TRACE[conn 2] plan after re-opt: Exchange └─ Table(test)
{connectTime=2022-06-09T18:05:44-07:00, connectionDb=test_db, query=SELECT * FROM test as of 'HEAD^1'}
2022-06-09T18:05:49-07:00 TRACE[conn 2] spooling result row [FLOAT64(1) FLOAT64(1) NULL]
{connectTime=2022-06-09T18:05:44-07:00, connectionDb=test_db, query=SELECT * FROM test as of 'HEAD^1'}
2022-06-09T18:05:49-07:00 TRACE[conn 2] spooling result row [FLOAT64(2) FLOAT64(2) NULL]
{connectTime=2022-06-09T18:05:44-07:00, connectionDb=test_db, query=SELECT * FROM test as of 'HEAD^1'}
2022-06-09T18:05:49-07:00 TRACE[conn 2] spooling result row [FLOAT64(3) FLOAT64(3) NULL]
{connectTime=2022-06-09T18:05:44-07:00, connectionDb=test_db, query=SELECT * FROM test as of 'HEAD^1'}
2022-06-09T18:05:49-07:00 TRACE[conn 2] committing transaction DoltTransaction
{connectTime=2022-06-09T18:05:44-07:00, connectionDb=test_db, query=SELECT * FROM test as of 'HEAD^1'}
2022-06-09T18:0

**Comment by timsehn** [comment]
Interestingly, Mysql client returns the correct result:
```
import mysql.connector
import sys


def main():
    user = 'root'
    port = 3306
    db = 'test_as_of'

    connection = mysql.connector.connect(user=user,
                                         host="127.0.0.1",
                                         port=port,
                                         database=db)

    query = "select * from test as of 'HEAD^' where pk=%s"

    cursor = connection.cursor()
    cursor.execute(query, [0])

    results = cursor.fetchall()
    print(results)

    cursor.close()

    connection.close()

    exit(0)

main()
```
When run:
```
dolt $ python test-mysql-prepareds.py 
[(0, 0)]
```
And our server logs are correct:
```
2022-06-09T18:36:52-07:00 TRACE [conn 4] beginning new transaction {connectTime=2022-06-09T18:36:52-07:00, connectionDb=test_as_of, query=set autocommit=0}
2022-06-09T18:36:52-07:00 TRACE [conn 4] spooling result row [] {connectTime=2022-06-09T18:36:52-07:00, connectionDb=test_as_of, query=set autocommit=0}
2022-06-09T18:36:52-07:00 DEBUG [conn 4] returning result &{[] 1 0  [[]] <nil>} {connectTime=2022-06-09T18:36:52-07:00, connectionDb=test_as_of, query=set autocommit=0}
2022-06-09T18:36:52-07:00 DEBUG [conn 4] Query took 0ms {connectTime=2022-06-09T18:36:52-07:00, connectionDb=test_as_of, query=set autocommit=0}
2022-06-09T18:36:52-07:00 DEBUG [conn 4] Starting query {connectTime=2022-06-09T18:36:52-07:00, connectionDb=test_as_of, query=select * from test as of 'HEAD^' where pk=0}
2022-06-09T18:36:52-07:00 TRACE [conn 4] beginning execution {connectTime=2022-06-09T18:36:52-07:00, connectionDb=test_as_of, query=select * from test as of 'HEAD^' where pk=0}
2022-06-09T18:36:52-07:00 TRACE [conn 4] spooling result row [INT32(0) INT32(0)] {connectTime=2022-06-09T18:36:52-07:00, connectionDb=test_as_of, query=select * from test as of 'HEAD^' where pk=0}
2022-06-09T18:36:52-07:00 DEBUG [conn 4] returning result &{[name:"pk" type:INT32 charset:33 name:"c1" type:INT32 charset:33] 1 0  [[INT32(0) INT32(0)]] <nil>} {connectTime=2022-06-09T18:36:52-07:00, connectionDb=test_as_of, query=select * from test as of 'HEAD^' where pk=0}
2022-06-09T18:36:52-07:00 DEBUG [conn 4] Query took 1ms {connectTime=2022-06-09T18:36:52-07:00, connectionDb=test_as_of, query=select * from test as of 'HEAD^' where pk=0}
```

**Comment by timsehn** [comment]
So something about the way MariaDB client is sending us prepareds is breaking us in this case. More debugging needed.

**Comment by timsehn** [comment]
Nevermind. I had to turn on Prepareds instead of just doing substitution on the client.
```
import mysql.connector
import sys


def main():
    user = 'root'
    port = 3306
    db = 'test_as_of'

    connection = mysql.connector.connect(user=user,
                                         host="127.0.0.1",
                                         port=port,
                                         database=db)

    query = "select * from test as of 'HEAD^' where pk=%s"

    cursor = connection.cursor(prepared=True)
    cursor.execute(query, [0,])

    results = cursor.fetchall()
    print(results)

    cursor.close()

    connection.close()

    exit(0)

main()
```
Produces:
```
dolt $ python test-mysql-prepareds.py 
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/mysql/connector/connection_cext.py", line 489, in cmd_stmt_execute
    prep_stmt.stmt_execute(*args)
_mysql_connector.MySQLInterfaceError: Error while executing statement: The number of columns in the result set differs from the number of bound buffers. You must reset the statement, rebind the result set columns, and execute the statement again

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/timsehn/liquidata/dolt/test-mysql-prepareds.py", line 29, in <module>
    main()
  File "/Users/timsehn/liquidata/dolt/test-mysql-prepareds.py", line 18, in main
    cursor.execute(query, [0,])
  File "/usr/local/lib/python3.9/site-packages/mysql/connector/cursor_cext.py", line 976, in execute
    res = self._cnx.cmd_stmt_execute(self._stmt, *params)
  File "/usr/local/lib/python3.9/site-packages/mysql/connector/connection_cext.py", line 491, in cmd_stmt_execute
    raise errors.InterfaceError(str(err))
mysql.connector.errors.InterfaceError: Error while executing statement: The number of columns in the result set differs from the number of bound buffers. You must reset the statement, rebind the result set columns, and execute the statement again
```
And server logs:
```
2022-06-09T18:46:04-07:00 DEBUG [conn 8] Starting query {connectTime=2022-06-09T18:46:04-07:00, connectionDb=test_as_of, query=select * from test as of 'HEAD^' where pk=?}
2022-06-09T18:46:04-07:00 TRACE [conn 8] beginning execution {connectTime=2022-06-09T18:46:04-07:00, connectionDb=test_as_of, query=select * from test as of 'HEAD^' where pk=?}
2022-06-09T18:46:04-07:00 TRACE [conn 8] optimizing prepared plan for query: select * from test as of 'HEAD^' where pk=? {connectTime=2022-06-09T18:46:04-07:00, connectionDb=test_as_of, query=select * from test as of 'HEAD^' where pk=?}
2022-06-09T18:46:04-07:00 TRACE [conn 8] plan before re-opt: Filter(test.pk = 0)
 └─ Projected table access on [pk c1]
     └─ Table(test)
 {connectTime=2022-06-09T18:46:04-07:00, connectionDb=test_as_of, query=select * from test as of 'HEAD^' where pk=?}
2022-06-09T18:46:04-07:00 TRACE [conn 8] plan after re-opt: IndexedTableAccess(test on [test.pk] with ranges: [{[0, 0]}]) {connectTime=2022-06-09T18:46:04-07:00, connectionDb=test_as_of, query=select * from test as of 'HEAD^' where pk=?}
2022-06-09T18:46:04-07:00 TRACE [conn 8] spooling result row [INT32(0) INT32(0) NULL] {connectTime=2022-06-09T18:46:04-07:00, connectionDb=test_as_of, query=select * from test as of 'HEAD^' where pk=?}
2022-06-09T18:46:04-07:00 DEBUG [conn 8] returning result &{[name:"pk" type:INT32 charset:33 name:"c1" type:INT32 charset:33 name:"c2" type:INT32 charset:33] 1 0  [[INT32(0) INT32(0) NULL]] <nil>} {connectTime=2022-06-09T18:46:04-07:00, connectionDb=test_as_of, query=select * from test as of 'HEAD^' where pk=?}
2022-06-09T18:46:04-07:00 DEBUG [conn 8] Query took 2ms {connectTime=2022-06-09T18:46:04-07:00, connectionDb=test_as_of, query=select * from test as of 'HEAD^' where pk=?}
```
So, Dolt is not handling this well.

**Comment by n8layman** [comment]
You all rock!

**Comment by timsehn** [comment]
We have a fix. We're working through merging and testing it. It has some conflicts with another change.

We're looking to have a release out for you by COB PST.

**Comment by n8layman** [comment]
No rush on our end. Thanks for such rapid and through attention to the issue!

**Comment by max-hoffman** [comment]
We merged a bug fix where we lost table versioning for prepared statements with non-bindvar `ASOF` clauses. Losing the table version appeared as a schema incompatibility in some of these client side libraries, and in others a silently ignored ASOF clause and incorrect results.

Again, the way the problem bubbled up seems to change between libraries. The fix resolves the errors for libraries I have tested. If there are follow-on issues just let us know in a new ticket and we will squash them as they come up.

We will issue a release tonight after everyone gets their PRs in.
